### PR TITLE
Add KokkosComm::irecv, improve unit testing output

### PR DESCRIPTION
**Pull request by cwpearson**
Add an interface to KokkosComm:Req to register a callable at wait(). Use that interface to invoke unpack operations (if needed) at wait(). Add packer_type to MpiArgs.

### CODE REVIEW

File: docs/api/core.rst [hunk 1]
```diff
@@ -101,10 +105,16 @@ Related Types
 
     .. cpp:function:: void KokkosComm::Req::wait()
 
-        Call MPI_Wait on the held MPI_Request and drop copies of any previous arguments to Req::keep_until_wait().
+        Call MPI_Wait on the held MPI_Request and drop copies of any previous arguments to Req::drop_at_wait().
```
Comment (masterleinad) [inline docs/api/core.rst]: How are these changes related to the title of the pull request? Would it make more sense to split them into a separate pull request?

Comment (masterleinad) [reply docs/api/core.rst]: Are you changing the name of `keep_until_wait` in this pull request to `drop_at_wait`? If so, should that be in another pull request?

File: src/KokkosComm.hpp [hunk 2]
```diff
@@ -38,6 +39,11 @@ void send(const ExecSpace &space, const SendView &sv, int dest, int tag,
   return Impl::send(space, sv, dest, tag, comm);
 }
 
+template <typename Recv, typename ExecSpace>
```
Comment (masterleinad) [inline src/KokkosComm.hpp]: What kind of object is `Recv` to be?

File: src/KokkosComm_request.hpp [hunk 3]
```diff
@@ -23,6 +23,22 @@
 namespace KokkosComm {
 
 class Req {
+  // a type-erased callable. Req uses these to attach callbacks to be executed
+  // at wait
+  struct InvokableHolderBase {
+    virtual ~InvokableHolderBase() {}
```
Comment (masterleinad) [inline src/KokkosComm_request.hpp]: Why not
```suggestion
    virtual ~InvokableHolderBase() = default;
```
?

File: src/KokkosComm_request.hpp [hunk 4]
```diff
@@ -23,6 +23,22 @@
...
+
+    virtual void operator()() = 0;
+  };
+  template <Invokable Fn>
+  struct InvokableHolder : InvokableHolderBase {
+    InvokableHolder(const Fn &f) : f_(f) {}
+
+    virtual void operator()() override { return f_(); }
```
Comment (masterleinad) [inline src/KokkosComm_request.hpp]:
```suggestion
    virtual void operator()() override { f_(); }
```
? The return type is `void`. Should the return type of `f_` to be `void` be enforced?

File: src/impl/KokkosComm_irecv.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+#include "KokkosComm_traits.hpp"
+
+// impl
+#include "KokkosComm_include_mpi.hpp"
+
+namespace KokkosComm::Impl {
+
+template <typename ExecSpace, typename RecvView, typename MpiArgs>
+struct IrecvUnpacker {
+  IrecvUnpacker(const ExecSpace &space, RecvView &rv, MpiArgs &args)
```
Comment (masterleinad) [inline src/impl/KokkosComm_irecv.hpp:31]: Where do you ensure that `RecvView` is a `Kokkos::View` or `mdspan`?
Where do you ensure that `ExecSpace` is a `Kokkos` execution space?

File: src/impl/KokkosComm_irecv.hpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    Args args = Packer::allocate_packed_for(space, "packed", rv);
+    space.fence();
+    MPI_Irecv(KCT::data_handle(args.view), args.count, args.datatype, src, tag,
+              comm, &req.mpi_req());
+    req.call_and_drop_at_wait(IrecvUnpacker{space, rv, args});
+
+  } else {
+    using RecvScalar = typename RecvView::value_type;
+    MPI_Irecv(KCT::data_handle(rv), KCT::span(rv), mpi_type_v<RecvScalar>, src,
```
Comment (masterleinad) [inline src/impl/KokkosComm_irecv.hpp:68]: How do you ensure that `rv` lives long enough? Do we care?

Comment (cwpearson) [reply src/impl/KokkosComm_irecv.hpp:68]: We could keep it alive or not. I will raise it in slack and see what people think.

Comment (cwpearson) [reply src/impl/KokkosComm_irecv.hpp:68]: Keep it alive!

File: docs/design.rst [hunk 7]
```diff
@@ -1,10 +1,24 @@
...
 
-"Immediate" functions (e.g. `isend`) return a `KokkosComm::Req`, which can be `wait()`-ed to block until the input view can be reused. `Req` also manages the lifetimes of any intermediate views needed for packing the data, releasing those views when `wait()` is complete.
+Asynchronous MPI operations use an ``MPI_Request``, which is a handle that can be used to refer to the operation later (e.g., ``MPI_Wait``).
+
+KokkosComm has an analogous concept, `KokkosComm::Req`.
+"Immediate" functions (e.g. `isend`) return a `KokkosComm::Req`, which can be `wait()`-ed to block until the input view can be reused.
+
+There are three consequences
+
+First, to ensure compatibility with MPI semantics, KokkosComm immediate functions will call the corresponding MPI function before they return.
```
Comment (masterleinad) [inline docs/design.rst]:
```suggestion
- First, to ensure compatibility with MPI semantics, KokkosComm immediate functions will call the corresponding MPI function before they return.
```

File: docs/design.rst [hunk 8]
```diff
@@ -1,10 +1,24 @@
...
+
+Second, the KokkosComm packing strategy may require that an intermediate view be allocated, and this view needs to have a lifetime at least as long as the communication.
```
Comment (masterleinad) [inline docs/design.rst]:
```suggestion
- Second, the KokkosComm packing strategy may require that an intermediate view be allocated, and this view needs to have a lifetime at least as long as the communication.
```

File: docs/design.rst [hunk 9]
```diff
@@ -1,10 +1,24 @@
...
+The ``KokkosComm::Req::keep_until_wait`` interface allows the `KokkosComm::Req` to hold those views until ``wait`` is called.
+
+Third, for asynchronous receive operations, the packing strategy may require that the buffer provided by the underlying MPI operation be further unpacked.
```
Comment (masterleinad) [inline docs/design.rst]:
```suggestion
- Third, for asynchronous receive operations, the packing strategy may require that the buffer provided by the underlying MPI operation be further unpacked.
```

File: src/impl/KokkosComm_irecv.hpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  RecvView rv_;
+  MpiArgs args_;
+};
+
+/* FIXME: If RecvView is a Kokkos view, it can be a const ref
+   same is true for an mdspan?
+*/
+template <typename RecvView, typename ExecSpace>
+Req irecv(const ExecSpace &space, RecvView &rv, int src, int tag,
+          MPI_Comm comm) {
```
Comment (masterleinad) [inline src/impl/KokkosComm_irecv.hpp]: This can be updated after merging https://github.com/cwpearson/kokkos-comm/pull/11.

**Comment by cwpearson** [comment]
Closed in favor of https://github.com/kokkos/kokkos-comm/pull/32, an identical PR from my fork
